### DISCUSSION TITLE
Line by line comparison

### DISCUSSION THREAD

**Discussion by LAPIII** [WinMerge/winmerge#1245] [Ideas]
Comparing line by line would be much more useful than auto aligning the differences.

**Comment by zhe-mazhe** [comment]
[foo.txt](https://github.com/WinMerge/winmerge/files/8520537/foo.txt)
[bar.txt](https://github.com/WinMerge/winmerge/files/8520540/bar.txt)

for example, I would like to compare files foo.txt and bar.txt line by line, because they are registers' values from the same address range of 0x000 to 0xFFF, before and after an event in my code.

**Reply by zhe-mazhe** [reply]
a workaround that I found is to change one file's EOL from CRLF to LF, and then _not_ to treat carriage return types as the same.

**Comment by meteorquake** [comment]
I suspect this is what I commonly require.
There is a common class of task where you have two lists, both sorted alphabetically (or can be sorted using the tool), and you just want gaps to be shown on one side to indicate an omission wherever there is an item on the other side.
To cater for non-alphabetical ones I guess you'd just make two dictionaries of each side's lines and use those to decide when an item has a match or not against the other side. When no match, a gap is inserted on the opposite side.
No doubt there are online options but it makes sense to provide the algorithm within Winmerge...
Cheers, David

**Comment by sdottaka** [comment]
First, recent versions of WinMerge add 'none' to the diff algorithm. This is a primitive comparison method with no automatic alignment.
I think this is probably what is desired here.
Specify as follows.

![image](https://user-images.githubusercontent.com/98126/213221001-419e9bb3-4604-4b53-9327-d122d7ea079c.png)

Also, plugins exist to sort and compare unsorted text. You can select the plugin as follows.
'
![image](https://user-images.githubusercontent.com/98126/213221450-6920da76-bbd3-4c6b-b0c0-cd0a69c330ca.png)

**Reply by zhe-mazhe** [reply]
I installed the latest version (2.16.24) and can verify that the Diff algorithm has indeed a new option of "none".
This new option has resolved my issue above, i.e., compare 2 files without shifting line(s).
Thank you!

**Comment by meteorquake** [comment]
Hello! I'll have a look again, but from my experience None as its name suggests does nothing (so you can view the originals as-is) and Sort just sorts (it doesn't calculate or display gaps for differences (i.e. the absences in this scenario) )... d

**Reply by sdottaka** [reply]
Maybe I don't understand exactly what you want, but "Moved block detection" is close to that?
![image](https://user-images.githubusercontent.com/98126/213318489-6ff02a5a-400e-4754-be4d-5791dec4bdd5.png)

**Comment by meteorquake** [comment]
This is an example tried with Moved Block Detection (if I understand its purpose right), which doesn't recognise 4 of the 6 items match and 2 differ (I'm not sure what a block entails but a moved-line comparison would just need to make a dictionary of lines with occurrence counts to perform its task) -
![a](https://user-images.githubusercontent.com/95745806/213682481-472650fe-8463-495c-99cd-d21843302489.png)

**Reply by sdottaka** [reply]
The reason david and michael are not considered moved is because there is no newline at the end of the file.
Try adding a newline like below.

![image](https://user-images.githubusercontent.com/98126/213721284-ca6c591e-ac38-46b5-a9ec-3377cea4c9f5.png)

**Reply by meteorquake** [reply]
Sorry I misread your comment (I'm a bit ill today!) but came to the conclusion you did - so maybe that needs fixing? For such purposes End of File should (I think) be treated like an End of Line...

**Reply by sdottaka** [reply]
This behavior is not erroneous because the difference in the end of the line is important.
However, it would be better to have an option to handle this well.
